### PR TITLE
Fix race condition between bot and user activity

### LINKED ISSUE #3431: First welcome message is delayed for 5 seconds
## Version

`4.10.0`

## Describe the bug

The first welcome message coming in is being delayed for 5 seconds. Probably due to recent accessibility fix that would delay message up to 5 seconds if their "reply-to-id" activity was not arrived yet.

## Steps to reproduce

1. Go to MockBot2
1. You should see "Hello and welcome!"

## Expected behavior

Instead, the "Hello and welcome!" arrive 5 seconds after the Web Socket packet has arrived.

## Additional context

[Bug]

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #3431.

### Fixed

-  Fixes [#3431](https://github.com/microsoft/BotFramework-WebChat/issues/3431). Race condition between the first bot activity and first user activity, should not cause the first bot activity to be delayed, by [@compulim](https://github.com/compulim) in PR [#3705](https://github.com/microsoft/BotFramework-WebChat/pull/3705)

## Description

A race condition between the first bot activity and first user activity will cause the bot activity to be delayed if the user activity shows up first.

## Design

When the bot activity (or any activity with "replyToId") has arrived, we will check if its precedence is already in the transcript or not. If it is not, we will wait for up to 5 seconds. This is for resolving an accessibility issue due to protocol/SDK deficiency.

But since the first bot activity will have a "replyToId" pointing to a bogus activity (a.k.a. never exists), thus, the bot activity will always subject to delay.

To mitigate this issue, we will check the transcript to see if the bot activity is the very first (i.e. with a bogus activity). If it is the very first, then, we allow it to show up without delay.

The problem lies in our transcript check; we check for all kinds of activities, including activities from both bot and user. Due to the race condition, the user's activity is already in the transcript. Thus, the bot activity is not the "first"; we don't allow the first bot activity to show up immediately.

The check's primary object is to check if the "replyToId" is a bogus ID or not. The check should be true if the transcript does not have **any bot activities**, instead of **any activities**.

## Specific Changes

- Updated the transcript check in `queueIncomingActivitySaga.js`, we should only check for bot's activities, not every activity
- Added tests to prevent regression

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] ~I have updated documentation~

## Review Checklist

> This section is for contributors to review your work.

-  [x] Accessibility reviewed (tab order, content readability, alt text, color contrast)
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] ~CSS styles reviewed (minimal rules, no `z-index`)~
-  [x] ~Documents reviewed (docs, samples, live demo)~
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] ~`package.json` and `package-lock.json` reviewed~
-  [x] ~Security reviewed (no data URIs, check for nonce leak)~
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -42,6 +42,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 -  Fixes [#3565](https://github.com/microsoft/BotFramework-WebChat/issues/3565). Allow strikethrough `<s>` on sanitize markdown, by [@corinagum](https://github.com/corinagum) in PR [#3646](https://github.com/microsoft/BotFramework-WebChat/pull/3646)
 -  Fixes [#3672](https://github.com/microsoft/BotFramework-WebChat/issues/3672). Center the icon of send box buttons vertically and horizontally, by [@compulim](https://github.com/compulim) in PR [#3673](https://github.com/microsoft/BotFramework-WebChat/pull/3673)
 -  Fixes [#3683](https://github.com/microsoft/BotFramework-WebChat/issues/3683). Activities should be acknowledged when user scrolls to bottom, by [@compulim](https://github.com/compulim) in PR [#3684](https://github.com/microsoft/BotFramework-WebChat/pull/3684)
+-  Fixes [#3431](https://github.com/microsoft/BotFramework-WebChat/issues/3431). Race condition between the first bot activity and first user activity, should not cause the first bot activity to be delayed, by [@compulim](https://github.com/compulim) in PR [#3705](https://github.com/microsoft/BotFramework-WebChat/pull/3705)
```
Comment (corinagum) [inline CHANGELOG.md:45]:
```suggestion
-  Fixes [#3431](https://github.com/microsoft/BotFramework-WebChat/issues/3431). Race condition between the first bot activity and first user activity, which should not cause the first bot activity to be delayed, by [@compulim](https://github.com/compulim) in PR [#3705](https://github.com/microsoft/BotFramework-WebChat/pull/3705)
```

File: packages/core/src/sagas/queueIncomingActivitySaga.js [hunk 2]
```diff
@@ -48,14 +54,15 @@ function* queueIncomingActivity({ userID }) {
     { payload: { activity } },
     initialActivities
   ) {
-    // This is for accessibility issue.
+    // This is for resolving an accessibility issue.
```
Comment (corinagum) [inline packages/core/src/sagas/queueIncomingActivitySaga.js:57]: nice!

File: packages/testharness/src/index.js [hunk 3]
```diff
@@ -10,6 +10,7 @@ import classNames from 'classnames';
 import createDeferred from 'p-defer-es5';
 import expect from 'expect';
 import lolex from 'lolex';
+import Observable from 'core-js/features/observable';
```
Comment (corinagum) [inline packages/testharness/src/index.js:13]: do we need this in the es5 bundle too, by any chance? or this is just for testing?

Comment (compulim) [reply packages/testharness/src/index.js:13]: It is just for the test harness. The new test need to use Observable.

File: __tests__/html/accessibility.delayActivity.withReplyToId.html [hunk 4]
```diff
@@ -30,6 +30,9 @@
         // Wait for "Connecting..." message to dismiss
         clock.tick(600);
 
+        // TODO: [P1] #3707 This line should be removed when the SDK removed the bogus replyToId.
```
Comment (corinagum) [inline __tests__/html/accessibility.delayActivity.withReplyToId.html:33]:
```suggestion
        // TODO: [P1] #3707 This line should be removed when the SDK removes the bogus replyToId.
```

**Comment by corinagum** [comment]
Might want to change 'bogus' to 'non-existent' - I don't know if this is a polycheck violation or not.

**Review by corinagum — APPROVED** [review]
LGTM pending a few comments.

**Pull request merged**